### PR TITLE
Optimization of Kernels that related to DeepLabv3+

### PR DESCRIPTION
**Pull request by cjld**
This PR optimize the DeepLab v3+ 3 times faster than before.
It contains:
* optimization of depthwise conv(forward, backward filter, backward input)
* optimization of reduce_mean and reduce_sum
* dilation support of depthwise conv

### CODE REVIEW

File: paddle/fluid/operators/math/depthwise_conv.cu [hunk 1]
```diff
@@ -20,149 +21,235 @@ namespace paddle {
...
+    const int output_channels, const int output_height, const int output_width,
+    const int input_channels, const int input_height, const int input_width,
+    const int filter_multiplier, const int filter_height,
+    const int filter_width, const int stride_height, const int stride_width,
+    const int padding_height, const int padding_width, const int dilate_height,
+    const int dilate_width, T* filter_grad_data) {
+  T s = 0;
+
+  int gbid = ((blockIdx.z * gridDim.y) + blockIdx.y) * gridDim.x + blockIdx.x;
+  // int gtid = threadIdx.y * blockDim.x + threadIdx.x;
```
Comment (qingqing01) [inline paddle/fluid/operators/math/depthwise_conv.cu]: Remove this line.

File: paddle/fluid/operators/cub_reduce.h [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+    return;
+  }
+  if (rank == 2 && reduce_rank == 1 && reduce_dim[0] == 1) {
+    ReduceKernel2D<Tx, Ty, ReduceOp, TransformOp,
+                   BlockDim><<<left_num, BlockDim, 0, stream>>>(
+        x_data, y_data, reducer, transformer, init, reduce_num);
+    return;
+  }
+  if (rank == 3 && reduce_rank == 1 && reduce_dim[0] == 1) {
+    // TODO(liangdun): we have to optimize 3d case which the 2nd axis is reduced
```
Comment (qingqing01) [inline paddle/fluid/operators/cub_reduce.h]: If not add this in PR, you can throw error:

```
PADDLE_THROW("xxx");
```

File: paddle/fluid/operators/math/depthwise_conv.cu [hunk 3]
```diff
@@ -20,149 +21,235 @@ namespace paddle {
...
+    for (int h_in = threadIdx.y; h_in < input_height; h_in += blockDim.y) {
+      const int batch = blockIdx.y;
+      const int c_in = blockIdx.x;
+
+      const int c_out_start = c_in * filter_multiplier;
+
+      int h_out_start =
+          (h_in - filter_height + padding_height + stride_height) /
+          stride_height;
+      int _h_out_start = 0 > h_out_start ? 0 : h_out_start;
```
Comment (qingqing01) [inline paddle/fluid/operators/math/depthwise_conv.cu]: h_out_start =  0 > h_out_start ? 0 : h_out_start;

Same as below, remove the fister letter:  `_`

File: paddle/fluid/operators/cub_reduce.h [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+      if ((is_reduced >> i) & 1)
+        new_reduce_dims.push_back(new_x_dim.size() - 1);
+    } else {
+      new_x_dim[new_x_dim.size() - 1] *= x_dim[i];
+    }
+  }
+  x_dim = new_x_dim;
+  // for (auto e : x_dim) VLOG(10) << "newdim " << e;
+  origin_reduce_dims = new_reduce_dims;
+  // for (auto e : origin_reduce_dims) VLOG(10) << "newrdim " << e;
```
Comment (qingqing01) [inline paddle/fluid/operators/cub_reduce.h]: Remove the commented code.

File: paddle/fluid/operators/cub_reduce.h [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+  int reduce_num = reduce_strides[0] * x_dim[reduce_dim[0]];
+  int left_num = 1;
+  if (left_dim.size()) left_num = left_strides[0] * x_dim[left_dim[0]];
+
+  std::vector<int> y_dim(left_dim.size());
+  for (int i = 0; i < left_dim.size(); ++i) {
+    y_dim[i] = x_dim[left_dim[i]];
+  }
+  auto x_data = x.data<Tx>();
+  // auto y_data = y->mutable_data<Ty>(framework::make_ddim(y_dim), x.place());
```
Comment (qingqing01) [inline paddle/fluid/operators/cub_reduce.h]: same as above.

File: paddle/fluid/operators/math/depthwise_conv.cu (hunk 3 above)
Comment (cjld) [reply paddle/fluid/operators/math/depthwise_conv.cu]: `_` is written on purpose, `_h_out_start` and `h_out_start` are not the same, see the code below

Comment (qingqing01) [reply paddle/fluid/operators/math/depthwise_conv.cu]: Ok. Just since `_h_out_start` doese follow Google C++ code style.

**Review by qingqing01 — DISMISSED** [review]

**Review by qingqing01 — APPROVED** [review]

**Pull request merged**